course/src/isometry.cpp

```diff
@@ -180,5 +180,181 @@ namespace math {
 
+  // Matrix --------------------------------------------
+
+  Matrix3::Matrix3(
+    const double a1, const double a2, const double a3,
+    const double b1, const double b2, const double b3,
+    const double c1, const double c2, const double c3) :
+    rows{Vector3{a1, a2, a3}, Vector3{b1, b2, b3}, Vector3{c1, c2, c3}} {}
+
+  Matrix3::Matrix3(
+    const Vector3& row1, const Vector3& row2, const Vector3& row3) :
+    rows{row1, row2, row3} { }
+
+  Matrix3::Matrix3() :
+    rows{
+      Vector3{0.0, 0.0, 0.0}, Vector3{0.0, 0.0, 0.0}, Vector3{0.0, 0.0, 0.0}} {}
+
+  const Matrix3 Matrix3::kIdentity =
+    Matrix3({1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0});
+
+  const Matrix3 Matrix3::kZero =
+    Matrix3({0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0});
+
+  const Matrix3 Matrix3::kOnes =
+    Matrix3({1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0});
+
+  bool Matrix3::operator==(const Matrix3& matrix1) const {
+    return(
+      rows.at(0) == matrix1.rows.at(0) &&
+      rows.at(1) == matrix1.rows.at(1) &&
+      rows.at(2) == matrix1.rows.at(2));
+  }
+
+  bool Matrix3::operator!=(const Matrix3& matrix1) const {
+    return(!(
+      rows.at(0) == matrix1.rows.at(0) &&
+      rows.at(1) == matrix1.rows.at(1) &&
+      rows.at(2) == matrix1.rows.at(2)));
+  }
+
+  Matrix3 Matrix3::operator+(const Matrix3& matrix1) const {
+    return {
+      rows.at(0)+matrix1.rows.at(0),
+      rows.at(1)+matrix1.rows.at(1),
+      rows.at(2)+matrix1.rows.at(2)};
+  }
+
+  Matrix3 Matrix3::operator-(const Matrix3& matrix1) const {
+    return {
+      rows.at(0)-matrix1.rows.at(0),
+      rows.at(1)-matrix1.rows.at(1),
+      rows.at(2)-matrix1.rows.at(2)};
+  }
+
+  Matrix3 Matrix3::operator*(const Matrix3& matrix1) const {
+    return {
+      rows.at(0)*matrix1.rows.at(0),
+      rows.at(1)*matrix1.rows.at(1),
+      rows.at(2)*matrix1.rows.at(2)};
+  }
+
+  Matrix3 Matrix3::operator/(const Matrix3& matrix1) const {
+    return {
+      rows.at(0)/matrix1.rows.at(0),
+      rows.at(1)/matrix1.rows.at(1),
+      rows.at(2)/matrix1.rows.at(2)};
+  }
+
+  Matrix3 Matrix3::operator/(const double divider) const {
+    return {
+      rows.at(0)/divider,
+      rows.at(1)/divider,
+      rows.at(2)/divider};
+  }
+
+  const Matrix3 operator*(const Matrix3& matrix1, const double scalar) {
+    return {
+      matrix1.rows.at(0)*scalar,
+      matrix1.rows.at(1)*scalar,
+      matrix1.rows.at(2)*scalar};
+  }
+
+  const Matrix3 operator*(const double scalar, const Matrix3& matrix1) {
+    return {
+      matrix1.rows.at(0)*scalar,
+      matrix1.rows.at(1)*scalar,
+      matrix1.rows.at(2)*scalar};
+  }
+
+  double Matrix3::det() const {
+    return
+      rows[0][0] * (rows[1][1]*rows[2][2] - rows[1][2]*rows[2][1]) -
+      rows[0][1] * (rows[1][0]*rows[2][2] - rows[1][2]*rows[2][0]) +
+      rows[0][2] * (rows[1][0]*rows[2][1] - rows[1][1]*rows[2][0]);
+  }
+
+  const Vector3 operator*(const Matrix3& matrix1, const Vector3& vector1) {
```

Comment:
Same as above, the return value should not be `const`.

Reply:
See e45c9eb